### PR TITLE
(#34) Fixes the starting point of the projectile trajectory

### LINKED ISSUE #34: Laser sight does not correspond to actual bullet trajectory when facing east
When you face west, the bullets follow the exact path that is expected by the Laser sight: [GIF](https://media.giphy.com/media/uT7aKtjhXiy6jES3Ro/giphy.gif)
But once you face east, the bullets stop following the path of the laser sight: [GIF](https://media.giphy.com/media/5SQitBLw9Dd5p6qeYp/giphy.gif)

It feels like the game makes the player switch hands to handle the gun, but it is not visually reflected. (or just game shouldn't make player switch hands)

### PR DESCRIPTION
**Pull request by y0ung3r**
image

Closes #34

### CODE REVIEW

**Comment by KeterSCP** [comment]
This change only fixes simple bullet (fired by `WLightChaingun` and `WMinigun`) starting point.

But there are other types of bullets in the game, some of them are rendered on the server (because their movement needs to be calculated on the server) like `PlasmaBall`

**Comment by y0ung3r** [comment]
@KeterSCP Thank you! I didn't know about these subtleties. I will translate the PR into a draft

**Comment by KeterSCP** [comment]
Seems like this pr also changes bullet's trajectories:

Main:

https://github.com/ForNeVeR/bloodmasters/assets/38229504/2b9248d4-0ab3-44c5-9a10-7dfbb5f29112

PR:

https://github.com/ForNeVeR/bloodmasters/assets/38229504/3a8f8c5a-7b06-4d50-994c-d2f3bee9f265

Notice how bullets missing laser end dot

**Comment by y0ung3r** [comment]
@KeterSCP If you pay attention to the angle at which the projectile flies out, it is clear that this problem is also present in the `main` branch. I didn't notice it last night when testing. But you are right, it shouldn't be

**Comment by y0ung3r** [comment]
@KeterSCP I'll try to fix it when I have some free time. Thank you!

**Comment by ForNeVeR** [comment]
> Seems like this pr also changes bullet's trajectories

JFYI, it is somewhat expected: see the discussion of #34 for more details.

**Comment by KeterSCP** [comment]
Yeah, I know. But as you may notice on the second attachment video, bullets are now missing laser sight dot when facing west, which is also incorrect. It would be perfect to align bullet trajectory with laser sight

**Comment by ForNeVeR** [comment]
> Yeah, I know. But as you may notice on the second attachment video, bullets are now missing laser sight dot when facing east, which is also incorrect. It would be perfect to align bullet trajectory with laser sight

Yep, on this I agree.

File: Source/Server/Weapons/WGrenadeLauncher.cs [hunk 1]
```diff
@@ -50,11 +50,8 @@ protected override void ShootOnce()
         // Extra Z boost
         vel.z += 0.5f;
 
-        // Move projectile somewhat forward
-        Vector3D pos = client.State.pos + Vector3D.FromActorAngle(client.AimAngle, client.AimAngleZ, PROJECTILE_OFFSET);
-
         // Spawn projectile
-        new Grenade(pos + new Vector3D(0f, 0f, PROJECTILE_Z), vel, client);
+        new Grenade(GetProjectileInitialPosition(client), vel, client);
```
Comment (ForNeVeR) [inline Source/Server/Weapons/WGrenadeLauncher.cs:54]: If we don't want to use `PROJECTILE_Z` anymore, let's delete that variable from all the affected weapons (and I assume from _all_ the weapons, that is?).

File: Source/Server/Weapons/WPhoenix.cs [hunk 2]
```diff
@@ -47,11 +47,8 @@ protected override void ShootOnce()
         // Determine projectile velocity
         Vector3D vel = Vector3D.FromActorAngle(client.AimAngle, client.AimAngleZ, PROJECTILE_VELOCITY);
 
-        // Move projectil somewhat forward
-        Vector3D pos = client.State.pos + Vector3D.FromActorAngle(client.AimAngle, client.AimAngleZ, PROJECTILE_OFFSET);
```
Comment (ForNeVeR) [inline Source/Server/Weapons/WPhoenix.cs:51]: Please note that `PROJECTILE_OFFSET` was different from `WPhoenix` (I mean it is equal to `4f` for everyone _except_ Phoenix) before this change.

Perhaps we should parametrize the `GetProjectileInitialPosition` method instead?

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
I am okay with the proposed solution, but let's consider using this `GetProjectileInitialPosition` when drawing the laser sight as well? If I understand correctly, then we'll have 100% correct laser sight, and all the projectiles will launch from the weapon origin and not the player head.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
See the comments.

**Comment by y0ung3r** [comment]
@ForNeVeR It still seems to me that there are a lot of potential problems in the proposed solution that are quite difficult to identify and verify. I think it is necessary to approach this problem from another side, but not in this PR anymore

**Pull request closed without merging**